Bugzilla/Config.pm

```diff
@@ -11,91 +11,123 @@ use 5.10.1;
-# when it shouldn't
 %Bugzilla::Config::EXPORT_TAGS
-  = (admin => [qw(update_params SetParam write_params)],);
+  = (admin => [qw(SetParam write_params)],);
 Exporter::export_ok_tags('admin');
 
-# INITIALIZATION CODE
-# Perl throws a warning if we use bz_locations() directly after do.
-our %params;
-
-# Load in the param definitions
-sub _load_params {
-  my $panels = param_panels();
-  my %hook_panels;
-  foreach my $panel (keys %$panels) {
-    my $module = $panels->{$panel};
-    require_module($module);
-    my @new_param_list = $module->get_param_list();
-    $hook_panels{lc($panel)} = {params => \@new_param_list};
-  }
+###############################
+####    Initialization     ####
+###############################
 
-  # This hook is also called in editparams.cgi. This call here is required
-  # to make SetParam work.
-  Bugzilla::Hook::process('config_modify_panels', {panels => \%hook_panels});
+sub new {
+  my $invocant = shift;
+  my $class    = ref($invocant) || $invocant;
 
-  foreach my $panel (keys %hook_panels) {
-    foreach my $item (@{$hook_panels{$panel}->{params}}) {
-      $params{$item->{'name'}} = $item;
-    }
+  my $cache = Bugzilla->request_cache;
+  if ($cache->{params_obj}) {
+    return $cache->{params_obj};
   }
-}
 
-# END INIT CODE
+  my $self = {params => {}};
+  bless($self, $class);
 
-# Subroutines go here
+  if (my $cached_params = Bugzilla->memcached->get_params()) {
+    $self->{params} = $cached_params;
+  }
+  else {
+    try {
+      my $dbh  = Bugzilla->dbh;
+      my $rows = $dbh->selectall_arrayref('SELECT name, value FROM params');
+      foreach my $row (@$rows) {
+        my ($name, $value) = @$row;
+        $self->{params}->{$name} = $value;
+      }
+    }
+    catch {
+      WARN("Database not yet available: $_");
 
-sub param_panels {
-  my $param_panels = {};
-  my $libpath      = bz_locations()->{'libpath'};
-  foreach my $item ((glob "$libpath/Bugzilla/Config/*.pm")) {
-    $item =~ m#/([^/]+)\.pm$#;
-    my $module = $1;
-    $param_panels->{$module} = "Bugzilla::Config::$module"
-      unless $module eq 'Common';
+      # Load defaults if database is unavailable
+      my $defs = $self->_load_param_defs();
+      foreach my $key (keys %{$defs}) {
+        $self->{params}->{$key} = $defs->{$key}->{default};
+      }
+    };
   }
 
-  # Now check for any hooked params
-  Bugzilla::Hook::process('config_add_panels', {panel_modules => $param_panels});
-  return $param_panels;
-}
+  $cache->{params_obj} = $self;
 
-sub SetParam {
-  my ($name, $value) = @_;
+  return $self;
+}
 
-  _load_params unless %params;
-  die "Unknown param $name" unless (exists $params{$name});
+###############################
+####       Setters         ####
+###############################
 
-  my $entry = $params{$name};
+sub set_param {
+  my ($self, $name, $value) = @_;
 
-  # sanity check the value
+  die "Unknown param $name" unless (exists $self->{params}->{$name});
 
+  # Sanity check the value
```

Comment:
Language nit: perhaps replace the terminology here with "validate the value"? I was recently asked to not use the terms "sanity" or "sane" to describe correct values.

Reply:
Will do.